File: 5. 이진 탐색/가사 검색.js
'use strict'

class Node {
  constructor() {
    this.end = false
    this.child = {}
    this.lenType = {}
  }
}

class Trie {
  constructor() {
    this.root = new Node()
  }

  insert(word) {
    let cur = this.root
    let cnt = 0
    const n = word.length
    for (let x of word) {
      if (cur.child[x] === undefined) {
        cur.child[x] = new Node()
      }
      cur = cur.child[x]
      cnt += 1
      if (cur.lenType[n - cnt]) cur.lenType[n - cnt] += 1
      else cur.lenType[n - cnt] = 1
    }
    cur.end = true
  }

  getCount(query) {
    let cur = this.root
    let cnt = query.length
    for (let x of query) {
      if (x === '?') break
      if (cur.child[x]) {
        cur = cur.child[x]
        cnt -= 1
      }
    }
    return cur.lenType[cnt] ? cur.lenType[cnt] : 0
  }
}

function solution(words, queries) {
  const trie_left = new Trie()
  const trie_right = new Trie()
  const ans = []

  for (let word of words) {
    trie_left.insert(word)
    trie_right.insert(word.split('').reverse())
  }
  for (let query of queries) {
    if (query[0] === '?') {
      ans.push(trie_right.getCount(query.split('').reverse().join('')))
    } else {
      ans.push(trie_left.getCount(query))
    }
  }
  return ans
}

// console.log(solution(["frodo", "front", "frost", "frozen", "frame", "kakao"], ["fro??", "????o", "fr???", "fro???", "pro?"]))
console.log(solution(["frodo", "front", "frost", "frozen", "frame", "kakao"], ["ka???"]))

File: codeEg/5. 이진 탐색/가사 검색.py
import bisect
def solution(words, queries):
    data = [[] for _ in range(100001)]    
    data_rev = [[] for _ in range(100001)]
    ans = []
    for word in words:
        data[len(word)].append(word)
        data_rev[len(word)].append(word[::-1])

    for i in range(100001):
        data[i].sort()
        data_rev[i].sort()

    for q in queries:
        if q[0] == '?':
            left = bisect.bisect_left(data_rev[len(q)], q.replace('?', 'a')[::-1])
            right = bisect.bisect_right(data_rev[len(q)], q.replace('?', 'z')[::-1])
        else:
            left = bisect.bisect_left(data[len(q)], q.replace('?', 'a'))
            right = bisect.bisect_right(data[len(q)], q.replace('?', 'z'))
        ans.append(right-left)
    return ans
            

print(solution(["frodo", "front", "frost", "frozen", "frame", "kakao"],
["fro??", "????o", "fr???", "fro???", "pro?"]))